src/sentry/static/sentry/app/views/dashboardsV2/widgetQueries.tsx

```diff
@@ -123,14 +125,51 @@ class WidgetQueries extends React.Component<Props, State> {
 
   componentDidUpdate(prevProps: Props) {
     const {selection, widget} = this.props;
+
+    // We do not fetch data whenever the query name changes.
+    const [prevWidgetQueryNames, prevWidgetQueries] = prevProps.widget.queries.reduce(
+      ([names, queries]: [string[], Omit<WidgetQuery, 'name'>[]], {name, ...rest}) => {
+        names.push(name);
+        queries.push(rest);
+        return [names, queries];
+      },
+      [[], []]
+    );
+
+    const [widgetQueryNames, widgetQueries] = widget.queries.reduce(
+      ([names, queries]: [string[], Omit<WidgetQuery, 'name'>[]], {name, ...rest}) => {
+        names.push(name);
+        queries.push(rest);
+        return [names, queries];
+      },
+      [[], []]
+    );
+
     if (
       !isEqual(widget.displayType, prevProps.widget.displayType) ||
       !isEqual(widget.interval, prevProps.widget.interval) ||
-      !isEqual(widget.queries, prevProps.widget.queries) ||
+      !isEqual(widgetQueries, prevWidgetQueries) ||
       !isEqual(widget.displayType, prevProps.widget.displayType) ||
       !isSelectionEqual(selection, prevProps.selection)
     ) {
       this.fetchData();
+      return;
+    }
+
+    if (
+      !isEqual(prevWidgetQueryNames, widgetQueryNames) &&
+      this.state.rawResults?.length === widget.queries.length
+    ) {
+      // If the query names has changed, then update timeseries labels
+
+      // eslint-disable-next-line react/no-did-update-set-state
+      this.setState(prevState => {
+        const timeseriesResults = widget.queries.reduce((acc: Series[], query, index) => {
+          return acc.concat(transformResult(query, prevState.rawResults![index]));
+        }, []);
```

Comment:
Could this act up if there are requests in flight due to condition changes? Sequence would be update conditions, and then quickly update the name before the request finishes?

Reply:
Indeed it could. I've restricted the condition to only update the labels whenever `state.loading` is `false`.